examples/chat/module_1/rust/src/lib.rs

```diff
@@ -14,78 +14,78 @@
-                info!("room terminating with {}", err.value());
-                for writer in &mut rsp_writers {
-                    writer.close(Ok(())).expect("Failed to close writer");
-                }
-                return err.value();
-            }
+        let in_channel = oak::ReadHandle {
+            handle: oak::Handle::from_raw(in_handle),
         };
-        if ready_status[0] != oak::ChannelReadStatus::READ_READY {
-            continue;
-        }
+        info!("starting event loop");
+        loop {
+            info!("waiting for new messages");
+            let ready_status = match oak::wait_on_channels(&[in_channel]) {
+                Ok(ready_status) => ready_status,
+                Err(err) => {
+                    // The other side of the channel was closed.
+                    info!("room terminating with {}", err.value());
+                    self.close_all();
+                    return err.value();
+                }
+            };
+            if ready_status[0] != oak::ChannelReadStatus::READ_READY {
+                continue;
+            };
 
-        // Read a message and possible additional outbound handles.
-        let mut buf = Vec::<u8>::with_capacity(512);
-        let mut handles = Vec::with_capacity(2);
-        let status = oak::channel_read(in_channel, &mut buf, &mut handles);
-        if status != oak::OakStatus::OK {
-            return status.value();
-        }
-        for handle in handles {
-            info!("add handle {:?} to output writers", handle);
-            rsp_writers.push(oak::grpc::ChannelResponseWriter::new(oak::WriteHandle {
-                handle,
-            }));
-        }
-        if buf.is_empty() {
-            continue;
-        }
-        let msg: ReceivedMessage = match protobuf::parse_from_bytes(&buf) {
-            Err(_) => {
-                warn!("failed to parse ReceivedMessage");
-                return oak::OakStatus::ERR_INTERNAL.value();
+            info!("reading incoming command");
+            let command: Command =
+                chat_common::receive(in_channel).expect("could not receive command");
+            match command {
+                Command::Join(h) => {
+                    self.clients.push(oak::grpc::ChannelResponseWriter::new(h));
+                }
+                Command::SendMessage(message_bytes) => {
+                    let message: Message = protobuf::parse_from_bytes(&message_bytes)
+                        .expect("could not parse message from bytes");
+                    self.messages.push(message.clone());
+                    info!("fan out message to {} clients", self.clients.len());
+                    for writer in &mut self.clients {
+                        // TODO: Improve error handling.
+                        writer
+                            .write(&message, oak::grpc::WriteMode::KeepOpen)
+                            .expect("could not write to channel");
+                    }
+                }
```

Comment:
@anghelcovici @daviddrysdale I am not sure it was apparent, but I think the main contribution of this PR is to encapsulate the core logic of the interaction between nodes in this `match` expression, and have algebraic data types (serializable enums containing handles in this case) drive the state changes. This is somewhat similar to https://guide.elm-lang.org/architecture/ and Redux, and also not too far from what happens behind the scenes in gRPC or FIDL when interacting using services or protocols. The point is that now almost all the rest of the code could be at least in principle generated, and developers would only write this command dispatching logic. This of course may be abstracted even more, but a layering is emerging that hopefully should make that easier in the future.